src/transformStateWithClones.js

```diff
@@ -7,7 +7,31 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  const stateHistory = [];
+  let currentState = { ...state };
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'addProperties':
+        currentState = { ...currentState, ...action.extraData };
+        break;
+
+      case 'removeProperties':
+        currentState = { ...currentState };
+
+        for (const key of action.keysToRemove) {
+          delete currentState[key];
+        }
+        break;
+
+      case 'clear':
+        currentState = {};
+        break;
+    }
+    stateHistory.push(currentState);
```

Comment:
The current implementation pushes `currentState` directly into `stateHistory`. Since `currentState` is an object, it is passed by reference, meaning that subsequent modifications to `currentState` will affect the objects already in `stateHistory`. To fix this, ensure that a new object is created and pushed to `stateHistory` after each action. You can achieve this by using `stateHistory.push({ ...currentState });` to create a shallow copy of `currentState`.